contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -1,9 +1,363 @@
+        return a > b ? a : b;
+    }
+
+    /**
+     * @dev Verify that an external signer signed the paymaster data of a user operation.
+     * The paymaster data is expected to be the paymaster and a signature over the entire request parameters.
+     * @param userOp The UserOperation struct that represents the current user operation.
+     * userOpHash The hash of the UserOperation struct.
+     * @param requiredPreFund The required amount of pre-funding for the paymaster.
+     * @return context A context string returned by the entry point after successful validation.
+     * @return validationData An integer returned by the entry point after successful validation.
+     */
+    function _validatePaymasterUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 requiredPreFund
+    ) internal override returns (bytes memory context, uint256 validationData) {
+        // review: in this method try to resolve stack too deep (though via-ir is good enough)
+        (
+            address paymasterId,
+            uint48 validUntil,
+            uint48 validAfter,
+            uint32 priceMarkup,
+            bytes calldata signature
+        ) = parsePaymasterAndData(userOp.paymasterAndData);
+
+        bytes32 hash = getHash(
+            userOp,
+            paymasterId,
+            validUntil,
+            validAfter,
+            priceMarkup
+        );
+        context = "";
+        // Review: var can be removed
+        uint256 sigLength = signature.length;
+        // we only "require" it here so that the revert reason on invalid signature will be of "VerifyingPaymaster", and not "ECDSA"
+        if (sigLength != 65) revert InvalidPaymasterSignatureLength(sigLength);
+        //don't revert on signature failure: return SIG_VALIDATION_FAILED
+        if (
+            verifyingSigner != hash.toEthSignedMessageHash().recover(signature)
+        ) {
+            // empty context and sigFailed with time range provided
+            return (context, _packValidationData(true, validUntil, validAfter));
+        }
+
+        require(
+            priceMarkup <= 2e6,
+            "Verifying PM: price markup percentage too high"
+        );
+        // Review if below is needed (as paymaster service may pass 0 if dynamic pricing doesn't apply)
+        require(
+            priceMarkup >= 1e6,
+            "Verifying PM: price markup percentage too low"
+        );
+
+        // Review: may not be needed at all
+        address account = userOp.getSender();
+
+        // Review max or min
+        uint32 dynamicMarkup = maxuint32(priceMarkup, fixedPriceMarkup);
+
+        if (
+            (requiredPreFund * dynamicMarkup) / PRICE_DENOMINATOR >
+            paymasterIdBalances[paymasterId]
+        )
+            revert InsufficientBalance(
+                (requiredPreFund * dynamicMarkup) / PRICE_DENOMINATOR,
+                paymasterIdBalances[paymasterId]
+            );
```

Comment:
I ran a test on some similar code and found out that we can save up to 2k gas by simply storing `(requiredPreFund * dynamicMarkup) / PRICE_DENOMINATOR` in a memoryvariable

Reply:
2k is huge. will change it

Reply:
done please check

Reply:
I think you forgot to push 🙂

Reply:
pushed now.